---
title: "Examining False Negatives"
author: Bridget Hegarty, James Riddell
date: 10-30-2022
output: html_notebook
---
This Rmarkdown file examines the false positives from the output of CheckV, DeepVirFinder, Kaiju,
VIBRANT, VirSorter, and VirSorter2 on multiple training sets of microbial DNA, 
primarily from NCBI. Created from fungal, viral, bacterial, archeael, protist,
and plasmid DNA sequences

Please reach out to James Riddell (riddell.26@buckeyemail.osu.edu) or
Bridget Hegarty (beh53@case.edu) regarding any issues, or open an issue on github.

```{r setup-library}
library(ggplot2)
library(plyr)
library(reshape2)
library(viridis)
library(tidyr)
library(dplyr)
library(readr)
library(data.table)
library(pROC)
```

Import the file that combines the results from each of the tools from running "combining_tool_output.Rmd":
```{r}
viruses <- read_tsv("../IntermediaryFiles/viral_tools_combined.tsv")
```

This section defines a viralness score "keep_score" based on the tool classifications. 
A final keep_score above 1 indicates we will keep that sequence and call it viral.

VIBRANT
    Quality == "Complete Circular": +1
    Quality == "High Quality Draft": +1
    Quality == "Medium Quality Draft": +1
    Quality == "Low Quality Draft" & provirus: +0.5

Virsorter2
    Viral >= 50: +0.5
    Viral >= 0.95: +0.5
    RNA >= 0.9: +1
    lavidaviridae >= 0.9: +1
    NCLDV >= 0.9: +1

Virsorter
    category ==  1,4: +1
    category == 2,5: +0.5

DeepVirFinder:
    Score >= 0.7: +0.5

Tuning - No Viral Signature:
    Kaiju_viral = "cellular organisms": -0.5
    If host_genes >50 and NOT provirus: -1 
    If viral_genes == 0 and host_genes >= 1: -1
    If 3*viral_genes <= host_genes and NOT provirus: -1
    If length > 50,000 and hallmark <=1: -1
    If length < 5000 and checkv completeness <= 75: -0.5

Tuning - Viral Signature:
    Kaiju_viral = "Viruses": +1
    If %unknown >= 75 and length < 50000: +0.5
    If %viral >= 50: +0.5
    Hallmark > 2: +1
    

This script produces visualizations of these combined viral scorings and
includes ecological metrics like alpha diversity.

You can decide which combination is appropriate for them and only need use the
tools appropriate for your data.

```{r getting_viral_set_1}
getting_viral_set_1 <- function(input_seqs,
                                include_vibrant=FALSE, 
                                include_virsorter2=FALSE,
                                include_deepvirfinder=FALSE,
                                include_tuning_viral=FALSE,
                                include_tuning_not_viral=FALSE,
                                include_virsorter=FALSE) {
  
  keep_score <- rep(0, nrow(input_seqs))
  
  if (include_vibrant) {
    keep_score[input_seqs$vibrant_quality=="complete circular"] <- keep_score[input_seqs$vibrant_quality=="complete circular"] + 1
    keep_score[input_seqs$vibrant_quality=="high quality draft"] <- keep_score[input_seqs$vibrant_quality=="high quality draft"] + 1
    keep_score[input_seqs$vibrant_quality=="medium quality draft"] <- keep_score[input_seqs$vibrant_quality=="medium quality draft"] + 1
    keep_score[input_seqs$vibrant_quality=="low quality draft" & input_seqs$provirus] <- keep_score[input_seqs$vibrant_quality=="low quality draft" & input_seqs$provirus] + 0.5
  }
  
  if (include_virsorter2) {
    keep_score[input_seqs$viral>=50] <- keep_score[input_seqs$viral>=50] + 0.5
    keep_score[input_seqs$viral>=95] <- keep_score[input_seqs$viral>=95] + 0.5
    keep_score[input_seqs$RNA>=0.9] <- keep_score[input_seqs$RNA>=0.9] + 1
    keep_score[input_seqs$lavidaviridae>=0.9] <- keep_score[input_seqs$lavidaviridae>=0.9] + 1
    keep_score[input_seqs$NCLDV>=0.9] <- keep_score[input_seqs$NCLDV>=0.9] + 1
  }
  
  if (include_virsorter) {
    keep_score[input_seqs$category==1] <- keep_score[input_seqs$category==1] + 1
    keep_score[input_seqs$category==2] <- keep_score[input_seqs$category==2] + 0.5
    keep_score[input_seqs$category==4] <- keep_score[input_seqs$category==4] + 1
    keep_score[input_seqs$category==5] <- keep_score[input_seqs$category==5] + 0.5
  }
  
  if (include_deepvirfinder) {
    keep_score[input_seqs$score>=0.7 & input_seqs$checkv_length<20000] <- keep_score[input_seqs$score>=0.7 & input_seqs$checkv_length<20000] + 0.5
    keep_score[input_seqs$score>=0.9 & input_seqs$checkv_length<20000] <- keep_score[input_seqs$score>=0.9 & input_seqs$checkv_length<20000] + 0.5
  }
  
  if (include_tuning_viral) {
    keep_score[input_seqs$Kaiju_Viral=="Viruses"] <- keep_score[input_seqs$Kaiju_Viral=="Viruses"] + 0.5
    keep_score[input_seqs$hallmark>2] <- keep_score[input_seqs$hallmark>2] + 1
    keep_score[input_seqs$percent_unknown>=75 & input_seqs$checkv_length<50000] <- keep_score[input_seqs$percent_unknown>=75 & input_seqs$checkv_length<50000] + 0.5
    keep_score[input_seqs$percent_viral>=50] <- keep_score[input_seqs$percent_viral>=50] + 0.5
  }
  
  if (include_tuning_not_viral) {
    keep_score[input_seqs$Kaiju_Viral=="cellular organisms"] <- keep_score[input_seqs$Kaiju_Viral=="cellular organisms"] - 0.5
    keep_score[input_seqs$checkv_host_genes>50 & !input_seqs$provirus] <- keep_score[input_seqs$checkv_host_genes>50 & !input_seqs$provirus] - 1
    keep_score[input_seqs$checkv_viral_genes==0 & input_seqs$checkv_host_genes>=1] <- keep_score[input_seqs$checkv_viral_genes==0 & input_seqs$checkv_host_genes>=1] - 1
    keep_score[((input_seqs$checkv_viral_genes*3) <= input_seqs$checkv_host_genes) & !input_seqs$provirus] <- keep_score[((input_seqs$checkv_viral_genes*3) <= input_seqs$checkv_host_genes) & !input_seqs$provirus] - 1 # consider accounting for RNA viruses
    keep_score[input_seqs$checkv_length>500000 & input_seqs$hallmark<=1] <- keep_score[input_seqs$checkv_length>500000 & input_seqs$hallmark<=1] - 1
    keep_score[input_seqs$checkv_completeness<=75 & input_seqs$checkv_length<=5000] <- keep_score[input_seqs$checkv_completeness<=75 & input_seqs$checkv_length<=5000] - 0.5 # helped with protist contamination
  }
  
  return(keep_score)
  
}
```


```{r}
viruses$keep_score_high_MCC <- getting_viral_set_1(viruses, include_deepvirfinder = F,
                                              include_vibrant = T,
                                              include_virsorter2 = T,
                                              include_tuning_viral = T,
                                              include_tuning_not_viral = T,
                                              include_virsorter = T)
```

```{r}
viruses$confusion_matrix_high_MCC <- "true negative"
viruses$confusion_matrix_high_MCC[viruses$seqtype=="virus" & viruses$keep_score_high_MCC<1] <- "false negative"
viruses$confusion_matrix_high_MCC[viruses$seqtype=="virus" & viruses$keep_score_high_MCC>=1] <- "true positive"
viruses$confusion_matrix_high_MCC[viruses$seqtype!="virus" & viruses$keep_score_high_MCC>=1] <- "false positive"
```

getting a list of false positive bacteria and plasmids from testing set 1
```{r}
fp_bacteria_list <- viruses$contig[viruses$Index==1 & 
                                     viruses$seqtype=="bacteria" &
                                     viruses$confusion_matrix_high_MCC=="false positive"]

fp_plasmid_list <- viruses$contig[viruses$Index==1 & 
                                     viruses$seqtype=="plasmid" &
                                     viruses$confusion_matrix_high_MCC=="false positive"]

tp_virus_list <- viruses$contig[viruses$Index==1 & 
                                     viruses$seqtype=="virus" &
                                     viruses$confusion_matrix_high_MCC=="true positive"]
```


vibrant annotations
```{r}
annotations_vb <- read_tsv("../ToolOutput/VIBRANT_annotations_metagenomic_testing_set_1.tsv")
```
```{r}
annotations_vb <- separate(annotations_vb, col="scaffold", into=c("seqtype", "contig"),
                           sep="--")
annotations_vb$contig <- sub("\\.", "_", annotations_vb$contig)
```


```{r}
head(annotations_vb)
```
```{r}
table(annotations_vb$contig %in% fp_bacteria_list)
table(annotations_vb$contig %in% fp_plasmid_list)
```

```{r}
annotations_vb_fp <- rbind(annotations_vb[annotations_vb$contig %in% fp_bacteria_list,],
                           annotations_vb[annotations_vb$contig %in% fp_plasmid_list,],
                           annotations_vb[annotations_vb$contig %in% tp_virus_list,])
```

```{r}
annotations_vb_fp_vog_count <- annotations_vb_fp %>%
  select(contig, VOG, seqtype) %>%
  group_by(contig, seqtype) %>%
  summarize(num_vog=table(is.na(VOG))[[1]],
            prop_vog=table(is.na(VOG))[[1]]/length(VOG),
            num_orfs=length(VOG))
```

```{r}
annotations_vb_fp_vog_count$confusion_matrix <- "false positive"
annotations_vb_fp_vog_count$confusion_matrix[annotations_vb_fp_vog_count$seqtype=="virus"] <- "true positive"
```


```{r}
pal <- ggthemes::tableau_color_pal(palette="Tableau 10", type="regular")
```

```{r}
ggplot(annotations_vb_fp_vog_count, 
       aes(x=seqtype, y=prop_vog, 
           color=seqtype, fill=seqtype)) +
  geom_boxplot() +
  geom_point(alpha=0.5) +
  theme_light() +
  theme(
    panel.grid.major.y = element_blank(),
    panel.border = element_blank(),
    axis.ticks.y = element_blank(),
    legend.position = "bottom",
    axis.text.y=element_text(size=14),
    axis.text.x=element_text(size=14),
    legend.text=element_text(size=12),
    axis.title=element_text(size=16),
  ) +
  ylab("Proportion Genes VOGs") +
  xlab("Sequence Type") +
  scale_fill_manual(name="",
                     values = alpha(rev(pal(3)), 0.5)) +
  scale_color_manual(name="",
                     values = alpha(rev(pal(3)), 1)) +
  facet_wrap(~confusion_matrix, scales="free_x")
```
```{r}
t.test(annotations_vb_fp_vog_count$prop_vog[annotations_vb_fp_vog_count$seqtype=="bacteria"],
       annotations_vb_fp_vog_count$prop_vog[annotations_vb_fp_vog_count$seqtype=="virus"])
```
```{r}
t.test(annotations_vb_fp_vog_count$prop_vog[annotations_vb_fp_vog_count$seqtype=="plasmid"],
       annotations_vb_fp_vog_count$prop_vog[annotations_vb_fp_vog_count$seqtype=="virus"])
```

```{r}
ggplot(annotations_vb_fp_vog_count, 
       aes(x=num_vog, y=prop_vog, 
           color=seqtype, fill=seqtype)) +
  geom_point(alpha=0.5) +
  theme_light() +
  theme(
    panel.grid.major.y = element_blank(),
    panel.border = element_blank(),
    axis.ticks.y = element_blank(),
    legend.position = "bottom",
    axis.text.y=element_text(size=14),
    axis.text.x=element_text(size=14),
    legend.text=element_text(size=12),
    axis.title=element_text(size=16),
  ) +
  ylab("Proportion Genes VOGs") +
  xlab("Number of VOGs") +
  scale_fill_manual(name="",
                     values = alpha(rev(pal(3)), 0.5)) +
  scale_color_manual(name="",
                     values = alpha(rev(pal(3)), 1))
```

# Thinking about short protists
```{r}
viruses[viruses$confusion_matrix_high_MCC=="false positive" & viruses$seqtype=="protist",]

table(viruses$confusion_matrix_high_MCC[viruses$seqtype=="protist"], 
      viruses$max_score_group[viruses$seqtype=="protist"])

table(viruses$Kingdom[viruses$seqtype=="protist" &  viruses$confusion_matrix_high_MCC=="false positive"], 
      viruses$max_score_group[viruses$seqtype=="protist" & viruses$confusion_matrix_high_MCC=="false positive"])

table(viruses$percent_viral[viruses$seqtype=="protist" &  viruses$confusion_matrix_high_MCC=="false positive"]>20, 
      viruses$max_score_group[viruses$seqtype=="protist" & viruses$confusion_matrix_high_MCC=="false positive"])
```

```{r}
ggplot(viruses, aes(y=percent_viral, x=confusion_matrix_high_MCC)) +
  geom_boxplot() +
  geom_point() +
  theme_bw() +
  facet_wrap(~seqtype, scales = "free") 
```
```{r}
table(viruses$confusion_matrix_high_MCC[viruses$checkv_length>=5000])
```







