---
title: "Comparing rule sets across environments"
output: html_notebook
---

```{r setup-library}
library(ggplot2)
library(plyr)
library(reshape2)
library(viridis)
library(tidyr)
library(dplyr)
library(readr)
library(data.table)
```


```{r}
viruses_dw <- read_csv("../DrinkingWaterMetagenomes/drinking_water_num_viruses.csv")
```
```{r}
viruses_rest <- read_csv("/Users/hegartyb/Downloads/num_viruses_env.csv")
```

only keeping chlorinated drinking water samples
```{r}
viruses_dw <- viruses_dw[viruses_dw$assembly %in% c("Birmingham_Maria",
                                                       "Cardiff_Maria", "Dundee_Clat",
                                                       "Dundee_Maria",
                                                       "Exeter_Maria", "Glasgow_Bal",
                                                       "Glasgow_Mil", "London_Maria"),
                           ]
```

```{r}
viruses_dw$environment <- "drinking water"

viruses_rest$environment[viruses_rest$environment=="waste_water"] <- "wastewater"
viruses_rest$environment[viruses_rest$environment=="lake_eerie"] <- "Lake Erie"
viruses_rest$environment[viruses_rest$environment=="lake_michigan"] <- "Lake Michigan"
viruses_rest$environment[viruses_rest$environment=="tara"] <- "global oceans"

viruses_all <- rbind(viruses_dw, viruses_rest)
```

recoding AJ's samples to be able to separate virome from metagenome (and remove others)
```{r}
viruses_all$environment2 <- viruses_all$environment
viruses_all$environment2[viruses_all$assembly %in% c("49639", "49640", "49641", "49642")] <- "Lake Erie virome"
viruses_all$environment2[viruses_all$assembly %in% c("49613", "49614", "49618", "49621",
                                                     "49624", "49628", "49632", "49638")] <- "Lake Erie metagenome"
viruses_all <- viruses_all[viruses_all$environment2!="Lake Erie",]
```




```{r}
viruses_all$environment2 <- factor(viruses_all$environment2,
                                    levels = c("drinking water", "Lake Erie metagenome", 
                                               "global oceans",
                                               "wastewater", "Lake Erie virome", 
                                               "Lake Michigan"))
```


```{r}
viruses_all$ruletype <- "other"
viruses_all$ruletype[viruses_all$toolcombo %in% c("0 0 1 0 0 1",
                                                          "0 0 1 1 0 0",
                                                          "0 0 0 1 0 0")] <- "high precision"
viruses_all$ruletype[viruses_all$toolcombo %in% c("1 1 0 0 1 1",
                                                          "1 1 0 1 1 1",
                                                          "1 1 0 0 0 1",
                                                          "1 1 0 1 0 1")] <- "high recall"
viruses_all$ruletype[viruses_all$toolcombo %in% c("0 0 1 0 0 1",
                                                          "0 1 1 0 0 1",
                                                          "0 0 0 0 0 1",
                                                          "1 0 1 1 0 1",
                                                          "1 0 1 0 1 1",
                                                          "1 0 1 0 0 1")] <- "high MCC"
viruses_all$ruletype[viruses_all$toolcombo %in% c("1 1 1 1 1 1")] <- "all"
```

```{r}
viruses_all$type <- "metagenome"
viruses_all$type[viruses_all$environment2 %in% c("wastewater", "Lake Erie virome",
                                                 "Lake Michigan")] <- "virome"
```


```{r}
ggplot(viruses_all, aes(x=toolcombo, y=prop_vir, 
                                  color=ruletype, fill=ruletype)) +
  geom_boxplot() +
  #geom_point() +
  theme_light() +
  theme(
    panel.grid.major.y = element_blank(),
    panel.border = element_blank(),
    axis.ticks.y = element_blank(),
    legend.position = "bottom",
    axis.text.y=element_text(size=14),
    axis.text.x=element_text(size=14, angle = 90),
    legend.text=element_text(size=12),
    axis.title=element_text(size=16),
    strip.background = element_rect(fill="white", color="grey"),
    strip.text = element_text(color="black")
  ) +
  xlab("Tool Combination (tnv, DVF, tv, VB, VS, VS2)") +
  ylab("Proportion of Contigs >3kb Viral") +
  scale_fill_manual(name="",
                     values = alpha(c(rev(magma(7)[3:6]), "grey"), 0.4)) +
  scale_color_manual(name="",
                     values = alpha(c(rev(magma(7)[3:6]), "grey"), 0.6)) +
  facet_grid(~environment2, scales = "free")
```



```{r}
viruses_all %>% 
  select(prop_vir, toolcombo, environment2) %>%
  filter(toolcombo=="1 0 0 1 0 0") %>%
  filter(environment2=="wastewater") 
  group_by(environment2) %>%
  summarise(mean = mean(prop_vir))
  
viruses_all %>% 
  select(prop_vir, toolcombo, environment2) %>%
  filter(toolcombo=="0 1 1 0 1 1") %>%
  group_by(environment2) %>%
  summarise(mean = mean(prop_vir))
  
viruses_all %>% 
  select(prop_vir, toolcombo, environment2) %>%
  filter(toolcombo=="1 0 0 0 0 1") %>%
  group_by(environment2) %>%
  summarise(mean = mean(prop_vir))
  
viruses_all %>% 
  select(prop_vir, toolcombo, environment2) %>%
  filter(toolcombo=="1 1 1 1 1 1") %>%
  group_by(environment2) %>%
  summarise(mean = mean(prop_vir))
```


```{r}
fig <- ggplot(viruses_all, aes(x=ruletype, y=prop_vir, 
                                  color=ruletype, fill=ruletype)) +
  geom_boxplot() +
  #geom_point() +
  theme_light() +
  theme(
    panel.grid.major.y = element_blank(),
    panel.border = element_blank(),
    axis.ticks.y = element_blank(),
    legend.position = "bottom",
    axis.text.y=element_text(size=10),
    axis.text.x=element_text(size=12, angle = 90, vjust = -0.1),
    legend.text=element_text(size=12),
    axis.title=element_text(size=12),
    strip.background = element_rect(fill="white", color="grey"),
    strip.text = element_text(color="black", size=12)
  ) +
  xlab("Environment") +
  ylab("Proportion of Contigs >3kb Viral") +
  scale_fill_manual(name="",
                     values = alpha(c(rev(magma(7)[3:6]), "grey"), 0.4)) +
  scale_color_manual(name="",
                     values = alpha(c(rev(magma(7)[3:6]), "grey"), 0.6)) +
  facet_wrap(~environment2)

fig

ggsave(
  "../IntermediaryFiles/simplified_environmental.png",
  plot = fig,
  scale = 1,
  width = 6.5,
  height = 4,
  units = c("in"),
  dpi = 300,
  limitsize = TRUE
)
```


```{r}
imp_toolcombs <- viruses_all %>% group_by(ruletype, environment2) %>%
  summarize(median=median(prop_vir),
            mean=mean(prop_vir),
            sd=sd(prop_vir)
            )
```

```{r}
ggplot(imp_toolcombs, aes(x=ruletype, y=mean, 
                                  color=ruletype, fill=ruletype)) +
  geom_point(size=3) +
  theme_light() +
  theme(
    panel.grid.major.y = element_blank(),
    panel.border = element_blank(),
    axis.ticks.y = element_blank(),
    legend.position = "bottom",
    axis.text.y=element_text(size=14),
    axis.text.x=element_text(size=14, angle = 90),
    legend.text=element_text(size=12),
    axis.title=element_text(size=16),
  ) +
  xlab("Tool Combination (tnv, DVF, tv, VB, VS, VS2)") +
  ylab("Proportion of Contigs >3kb Viral") +
  scale_fill_manual(name="",
                     values = alpha(rev(viridis(6)), 0.3)) +
  scale_color_manual(name="",
                     values = alpha(rev(viridis(6)), 0.7)) +
  facet_wrap(~environment2)
```





