---
title: "viral_contig_identification_user"
author: "James Riddell"
date: "8/25/2023"
output: html_document
---

This Rmarkdown file takes outputs from viral identification tools and merges
them into one file, "{dataset}_merged_viral_id_outputs.csv", which is the input
for identifying_viral_metrics_environmental.Rmd.

Please reach out to James Riddell (riddell.26@buckeyemail.osu.edu) or
Bridget Hegarty (beh53@case.edu) regarding any issues, or open an issue on github.


```{r setup-library}
library(ggplot2)
library(plyr)
library(reshape2)
library(viridis)
library(tidyr)
library(dplyr)
library(readr)
library(data.table)
```

# Inputs
1) For this script's inputs, the user must merge each viral identification 
tool output with the following specified filenames into one tab-separated file. 
Replace the filenames with their paths if needed into the variables in the 
'inputs' chunk.

    General output filenames used from each tool:
        Checkv: quality_summary.tsv
        VIBRANT: VIBRANT_genome_quality_${assembly}.tsv
        Virsorter: VIRSorter_global-phage-signal.csv
        Virsorter2: final-viral-score.tsv
        DeepVirFinder: ${assembly}.fasta_gt2500bp_dvfpred.txt
        Kaiju: ${assembly}.kaiju.names.out

2) Make sure all fasta headers (assembly, contig) are consistent within and
across tools. Chunks for each tool do contain some lines for cleaning up these features, but due to their variability it will be the user's responsibility to 
make sure they match across tools.

3) Check each chunk and ensure all columns are accounted for. For example,
if you don't have an "assembly" column and called it something else, please either
remove it from this script or modify your input files.

4) This script is designed for contigs > 3000 basepairs. It can be modified
to be higher or lower, but going lower will greatly increase the size of the
dataframe and memory usage.



```{r inputs}
# dataset name for file organization and outputs for metrics file
dataset_name <- "drinking_water"
# checkV
checkV_path <- "DrinkingWaterToolOutput/merged_checkV_allcontigs.tsv"
# VIBRANT
vibrant_path <- "DrinkingWaterToolOutput/vibrant_merged_contig_list_all.txt"
# DeepVirFinder
dvf_path <- "DrinkingWaterToolOutput/dvf_outputs_merged.tsv"
# Virsorter
vs_path <- "DrinkingWaterToolOutput/vs_merged_all.txt"
# Virsorter2
vs2_path <- "DrinkingWaterToolOutput/vs2_merged_file.txt"
# Kaiju
kj_path <- "/Users/hegartyb/Documents/BlueSkyPostdoc/ViralSortingToolsEvaluation/Temp/kaiju_outputs_merged.tsv"
# length cutoff (unit: basepairs)
BP_CUTOFF <- 3000

```

Here we will read in all the viral ID output files to 1) standardize their
column names, and 2) standardize the contig names. The contig name will act
as the key to join all of the outputs into one dataframe, so make sure it is
the same across tool outputs.

If you are not using all of the tool outputs, there is no need to run the
code chunks for tools you are not using.

## checkV
```{r}
checkV <- fread(checkV_path, 
                sep="\t",
                header = T, 
                select = c(
                    'contig_id',
                    'assembly',
                    'provirus',
                    'completeness',
                    'contamination',
                    'viral_genes',
                    'host_genes',
                    'gene_count',
                    'contig_length',
                    'checkv_quality'
                    )
                ) %>% 
    rename(
        contig = contig_id,
        checkv_provirus = provirus,
        checkv_completeness = completeness,
        checkv_contamination = contamination,
        checkv_viral_genes = viral_genes,
        checkv_host_genes = host_genes,
        checkv_total_genes = gene_count,
        checkv_length = contig_length
        )
checkV$method = 'checkv'
checkV$contig <- sub("\\.", "_", checkV$contig)
checkV$uniq_contig <- paste(checkV$assembly, checkV$contig, sep = "--")
checkV <- checkV[!duplicated(checkV$uniq_contig),]
```

## VIBRANT
```{r}
vb_c <- fread(vibrant_path,
              header = T,
              sep = "\t",
              select = c(
                  'scaffold',
                  'assembly',
                  'type',
                  'quality'
                  )
              ) %>%
    rename(
        contig = scaffold,
        vibrant_quality = quality
        )
vb_c$method <- "vibrant"
vb_c$vibrant_prophage <- "No"
vb_c$vibrant_prophage[grep("_fragment_", vb_c$contig)] <- "Yes"
vb_c$contig <- gsub("_fragment_.*", "", vb_c$contig)
vb_c$contig <- sub("\\.", "_", vb_c$contig)
vb_c$uniq_contig <- paste(vb_c$assembly, vb_c$contig, sep = "--")
vb_c <- vb_c[!duplicated(vb_c$uniq_contig),]
```

## DeepVirFinder
```{r}
dvf_c <- fread(dvf_path,
               header = T,
               sep = "\t",
               select = c(
                   'sample',
                   'name',
                   'score',
                   'pvalue'
                   )
               ) %>% 
    rename(
        contig = name,
        assembly = sample
        )
dvf_c$contig <- sub("\\.", "_", dvf_c$contig)
dvf_c$bh_pvalue <- p.adjust(dvf_c$pvalue, method = "BH")
dvf_c$uniq_contig <- paste(dvf_c$assembly, dvf_c$contig, sep = "--")
dvf_c <- dvf_c[!duplicated(dvf_c$uniq_contig),]
```

## VirSorter
```{r}
vs_c <- fread(vs_path,
              select = c(
                  'assembly',
                  'contig_id',
                  'category'
                  ),
              ) %>% 
  rename(contig = contig_id) %>%
  drop_na(contig)
vs_c$contig <- sub("VIRSorter_", "", vs_c$contig)
vs_c$contig <- sub("-circula", "", vs_c$contig)
vs_c$contig <- sub("uth", "South", vs_c$contig)
vs_c$contig <- sub("pannenburg", "Spannenburg", vs_c$contig)
vs_c$contig <- sub("\\.", "_", vs_c$contig)
vs_c$uniq_contig <- paste(vs_c$assembly, vs_c$contig, sep = "--")

vs_c <- vs_c[!duplicated(vs_c$uniq_contig),]
vs_c <- vs_c %>% drop_na(uniq_contig)
```


## VirSorter2 
```{r}
vs2_c <- fread(vs2_path,
                    header = T,
                    sep = '\t',
                    select = c(
                        'seqname',
                        'assembly',
                        'dsDNAphage',
                        'ssDNA',
                        'max_score',
                        'max_score_group',
                        'hallmark',
                        'viral',
                        'cellular'
                    )
               ) %>% 
              separate(
                  col = seqname,
                  into = c("contig", "type"), 
                  sep = "\\|\\|",
                  remove = T
                  )
vs2_c$assembly <- sub("_3000", "", vs2_c$assembly)
vs2_c$contig <- sub("\\.", "_", vs2_c$contig)
vs2_c$uniq_contig <- paste(vs2_c$assembly, vs2_c$contig, sep="--")
vs2_c <- vs2_c[!duplicated(vs2_c$uniq_contig),]
```


## Kaiju
```{r}
kj_c <- read_tsv(kj_path, col_names = T)
colnames(kj_c)[1] <- "assembly"
colnames(kj_c)[3] <- "contig"
kj_c$contig <- sub("\\.", "_", kj_c$contig)
kj_c <- separate(kj_c, col = Name, into = c("Kaiju_Viral","Kingdom"), sep=";")
kj_c$uniq_contig <- paste(kj_c$assembly, kj_c$contig, sep="--")
kj_c <- kj_c[!duplicated(kj_c$contig),]
```

Now we will merge all of the outputs together into one dataframe, 'viruses'.
If you are not using all of the tools, then there is no need to run the lines
below for the tools you are not using.

## Merging
```{r}
viruses <- full_join(x=checkV, y=vs2_c, by = c("uniq_contig", "assembly"))
viruses <- full_join(x=viruses, y=dvf_c, by = c("uniq_contig", "assembly"))
viruses <- full_join(x=viruses, y=vb_c, by = c("uniq_contig", "assembly"))
viruses <- full_join(x=viruses, y=vs_c, by = c("uniq_contig", "assembly"))
viruses <- full_join(x=viruses, y=kj_c, by = c("uniq_contig", "assembly"))
```

Remove contigs not greater than the basepair length cutoff. If you prefer
faster code, you can also move this step to the beginning with checkV, and
perform left joins instead of full joins to the checkV dataframe, assuming
all contigs are represented in the CheckV dataframe.

```{r}
viruses <- viruses %>% filter(checkv_length > BP_CUTOFF)
```


Use the next code chunk to check if any contigs are missing from checkv.
If there are any rows present, these are located in other viral ID outputs,
but not CheckV. Please go back and change the contig names so that they match
the pattern in CheckV, or check your CheckV output to make sure it contains
as many rows as headers in your fasta file (with the bp cutoff).
```{r}
v_missing <- viruses[is.na(viruses$checkv_uniq_contig),]
```



Calculate percent viral/host/unknown/kaiju_match_ratio columns
These will be used in the classification step.
```{r}
viruses$percent_host <- viruses$checkv_host_genes/viruses$checkv_total_genes*100
viruses$percent_viral <- viruses$checkv_viral_genes/viruses$checkv_total_genes*100
viruses$percent_unknown <- 100-(viruses$checkv_host_genes+viruses$checkv_viral_genes)/viruses$checkv_total_genes*100
viruses$kaiju_match_ratio <- viruses$len/viruses$checkv_length
```

get rid of NAs for downstream processing
```{r}

# checkV
viruses$percent_viral[is.na(viruses$percent_viral)] <- 0
viruses$percent_unknown[is.na(viruses$percent_unknown)] <- 0
viruses$checkv_completeness[is.na(viruses$checkv_completeness)] <- 0

# VIBRANT
viruses$vibrant_quality[is.na(viruses$vibrant_quality)] <- 0

# DeepVirFinder
viruses$score[is.na(viruses$score)] <- 0
viruses$bh_pvalue[is.na(viruses$bh_pvalue)] <- 0

# VirSorter2
viruses$viral[is.na(viruses$viral)] <- 0
viruses$hallmark[is.na(viruses$hallmark)] <- 0

#Virsorter
viruses$category[is.na(viruses$category)] <- 0

# Kaiju
viruses$Kaiju_Viral[is.na(viruses$Kaiju_Viral)] <- "unknown"
```

```{r}
viruses_filename <- 'merged_viral_feature_outputs.csv'
write.csv(viruses, paste(dataset_name, viruses_filename, sep='_'))
```

Data cleaning is done! Export viruses to tsv and load into identifying_viral_metrics.Rmd for metrics on tool outputs. The code for metrics is also pasted below, but it is not recommended running it in the same R session due to memory issues. The construction of the final dataframe essentially doubles the memory being used, and even with garbage collector can make Rstudio unresponsive.