---
title: "Viral Sequence Sorting A2"
output: html_notebook
---

based on VSTE v 20220524 

```{r}
library(ggplot2)
library(plyr)
library(reshape2)
library(viridis)
library(tidyr)
library(dplyr)
library(readr)
```

# All Viral Contigs

## checkV
```{r}
checkV <- read_tsv("../Results/contamination.tsv")
checkV$contig <- sub("\\.", "_", checkV$contig_id)
```

## VIBRANT
```{r}
vb_c <- read_tsv("../Results/VIBRANT_genome_quality_contigs_3000.tsv")
vb_c$contig <- sub("\\.", "_", vb_c$scaffold)
vb_c$contig <- gsub("_fragment_.*", "", vb_c$contig)
vb_c <- vb_c[!duplicated(vb_c$contig),]
```

## DeepVirFinder
```{r}
dvf_c <- read_tsv("../Results/contigs_3000.fasta_gt1bp_dvfpred.txt")
dvf_c$contig <- sub("\\.", "_", dvf_c$name)
dvf_c$bh_pvalue <- p.adjust(dvf_c$pvalue, method="BH")
```

## VirSorter2 
```{r}
vs2_c <- read_tsv("../Results/merged_final_viral_score_A2.tsv")
vs2_c <- separate(vs2_c, col = seqname, into = c("contig", "vs2type"), "\\|\\|")
vs2_c$contig <- sub("\\.", "_", vs2_c$contig)
vs2_c <- vs2_c[!duplicated(vs2_c$contig),]
```


## Kaiju
(parsing when reading the file in "fails" for sequences that weren't classified)
```{r}
kj_c <- read_tsv("../Results/A2.nr_euk.kaiju.names.out", col_names = F)
colnames(kj_c) <- c("Classified", "Contig", "NCBI_taxon", "len", "ID_best", "IDs_all", "Seq", "Name")
kj_c <- separate(kj_c, col = Name, into = c("Kaiju_Viral","Kingdom"), sep=";")
kj_c$contig <- sub("\\.", "_", kj_c$Contig)
```

```{r}
table(kj_c$seqtype, kj_c$Kaiju_Viral)
```

## Merging

```{r}
viruses <- full_join(x=vs2_c, y=dvf_c, by = c("contig"))
viruses <- full_join(x=viruses, y=checkV, by = c("contig"))
viruses <- full_join(x=viruses, y=vb_c, by = c("contig"))
viruses <- full_join(x=viruses, y=kj_c, by = c("contig"))
```

calculate percent viral/host/unknown
```{r}
viruses$percent_host <- viruses$host_genes/viruses$total_genes*100
viruses$percent_viral <- viruses$viral_genes/viruses$total_genes*100
viruses$percent_unknown <- 100-(viruses$host_genes+viruses$viral_genes)/viruses$total_genes*100
```

get rid of NAs for downstream processing
```{r}
viruses$Quality[is.na(viruses$Quality)] <- 0
viruses$score[is.na(viruses$score)] <- 0
viruses$bh_pvalue[is.na(viruses$bh_pvalue)] <- 0
viruses$viral[is.na(viruses$viral)] <- 0
viruses$hallmark[is.na(viruses$hallmark)] <- 0
viruses$percent_viral[is.na(viruses$percent_viral)] <- 0
viruses$percent_unknown[is.na(viruses$percent_unknown)] <- 0
viruses$Kaiju_Viral[is.na(viruses$Kaiju_Viral)] <- "unknown"
viruses$Kingdom[is.na(viruses$Kingdom)] <- "unknown"
```



##################################################


```{r}
getting_viral_set_1 <- function(input_seqs, include_vibrant=FALSE, include_virsorter2=FALSE,
                              include_deepvirfinder=FALSE, include_checkV=FALSE,
                              include_kaiju=FALSE) {
  
  keep_score <- rep(0, nrow(input_seqs))
  
  if (include_vibrant) {
    keep_score[input_seqs$Quality=="high quality draft"] <- keep_score[input_seqs$Quality=="high quality draft"] + 1
    keep_score[input_seqs$Quality=="medium quality draft"] <- keep_score[input_seqs$Quality=="medium quality draft"] + 1
    keep_score[input_seqs$Quality=="low quality draft" & input_seqs$provirus=="Yes"] <- keep_score[input_seqs$Quality=="low quality draft" & input_seqs$provirus=="Yes"] + 0.5
  }
  
  if (include_virsorter2) {
    keep_score[input_seqs$viral>=50] <- keep_score[input_seqs$viral>=50] + 0.5
    keep_score[input_seqs$hallmark>2] <- keep_score[input_seqs$hallmark>2] + 1
    keep_score[input_seqs$viral>=95] <- keep_score[input_seqs$viral>=95] + 1
  }
  
  if (include_deepvirfinder) {
    keep_score[input_seqs$score>=0.7] <- keep_score[input_seqs$score>=0.7] + 0.5
    keep_score[input_seqs$score>=0.9] <- keep_score[input_seqs$score>=0.9] + 0.5
  }
  
  if (include_kaiju) {
    keep_score[input_seqs$Kaiju_Viral=="cellular organisms"] <- keep_score[input_seqs$Kaiju_Viral=="cellular organisms"] - 1
    keep_score[input_seqs$Kaiju_Viral=="Viruses"] <- keep_score[input_seqs$Kaiju_Viral=="Viruses"] + 1
  }
  
  if (include_checkV) {
    keep_score[input_seqs$viral_genes==0 & input_seqs$host_genes>=1] <- 0
    keep_score[(input_seqs$viral_genes*3) <= input_seqs$host_genes] <- 0
    keep_score[input_seqs$contig_length>50000 & input_seqs$hallmark==0] <- 0
    keep_score[input_seqs$percent_unknown>=75] <- keep_score[input_seqs$percent_unknown>=75] + 0.5
  }
  
  return(keep_score)
  
}


```




## just vibrant
```{r}
viruses$keep_score_vb <- getting_viral_set_1(viruses, include_vibrant = T,
                                            include_virsorter2 = F,
                                            include_checkV = F,
                                            include_kaiju = F,
                                            include_deepvirfinder = F)
```

## vibrant + checkV
```{r}
viruses$keep_score_vb_cv <- getting_viral_set_1(viruses, include_vibrant = T,
                                            include_virsorter2 = F,
                                            include_checkV = T,
                                            include_kaiju = F,
                                            include_deepvirfinder = F)
```

## vibrant + checkV + kaiju
```{r}
viruses$keep_score_vb_cv_kj <- getting_viral_set_1(viruses, include_vibrant = T,
                                            include_virsorter2 = F,
                                            include_checkV = T,
                                            include_kaiju = T,
                                            include_deepvirfinder = F)
```

## vibrant + checkV + kaiju + deepvirfinder
```{r}
viruses$keep_score_vb_cv_kj_dvf <- getting_viral_set_1(viruses, include_vibrant = T,
                                            include_virsorter2 = F,
                                            include_checkV = T,
                                            include_kaiju = T,
                                            include_deepvirfinder = T)
```

## vibrant + virsorter2 + checkV + kaiju
```{r}
viruses$keep_score_vb_vs2_cv_kj <- getting_viral_set_1(viruses, include_vibrant = T,
                                            include_virsorter2 = T,
                                            include_checkV = T,
                                            include_kaiju = T,
                                            include_deepvirfinder = F)
```

#all: vibrant + virsorter2 + checkV + kaiju + deepvirfinder
```{r}
viruses$keep_score_all <- getting_viral_set_1(viruses, include_vibrant = T,
                                            include_virsorter2 = T,
                                            include_checkV = T,
                                            include_kaiju = T,
                                            include_deepvirfinder = T)
```

## just virsorter2
```{r}
viruses$keep_score_vs2 <- getting_viral_set_1(viruses, include_vibrant = F,
                                            include_virsorter2 = T,
                                            include_checkV = F,
                                            include_kaiju = F,
                                            include_deepvirfinder = F)
```

## virsorter2 + checkV
```{r}
viruses$keep_score_vs2_cv <- getting_viral_set_1(viruses, include_vibrant = F,
                                            include_virsorter2 = T,
                                            include_checkV = T,
                                            include_kaiju = F,
                                            include_deepvirfinder = F)
```

## virsorter2 + checkV + kaiju
```{r}
viruses$keep_score_vs2_cv_kj <- getting_viral_set_1(viruses, include_vibrant = F,
                                            include_virsorter2 = T,
                                            include_checkV = T,
                                            include_kaiju = T,
                                            include_deepvirfinder = F)
```

## virsorter2 + checkV + kaiju + deepvirfinder
```{r}
viruses$keep_score_vs2_cv_kj_dvf <- getting_viral_set_1(viruses, include_vibrant = F,
                                            include_virsorter2 = T,
                                            include_checkV = T,
                                            include_kaiju = T,
                                            include_deepvirfinder = T)
```

## just checkV
```{r}
viruses$keep_score_cv <- getting_viral_set_1(viruses, include_vibrant = F,
                                            include_virsorter2 = F,
                                            include_checkV = T,
                                            include_kaiju = F,
                                            include_deepvirfinder = F)
```

## just kaiju
```{r}
viruses$keep_score_kj <- getting_viral_set_1(viruses, include_vibrant = F,
                                            include_virsorter2 = F,
                                            include_checkV = F,
                                            include_kaiju = T,
                                            include_deepvirfinder = F)
```

## checkV + kaiju
```{r}
viruses$keep_score_kj_cv <- getting_viral_set_1(viruses, include_vibrant = F,
                                            include_virsorter2 = F,
                                            include_checkV = T,
                                            include_kaiju = T,
                                            include_deepvirfinder = F)
```


## just deepvirfinder
```{r}
viruses$keep_score_dvf <- getting_viral_set_1(viruses, include_vibrant = F,
                                            include_virsorter2 = F,
                                            include_checkV = F,
                                            include_kaiju = F,
                                            include_deepvirfinder = T)
```

## deepvirfinder + checkV
```{r}
viruses$keep_score_dvf_cv <- getting_viral_set_1(viruses, include_vibrant = F,
                                            include_virsorter2 = F,
                                            include_checkV = T,
                                            include_kaiju = F,
                                            include_deepvirfinder = T)
```

## deepvirfinder + checkV + kaiju
```{r}
viruses$keep_score_dvf_cv_kj <- getting_viral_set_1(viruses, include_vibrant = F,
                                            include_virsorter2 = F,
                                            include_checkV = T,
                                            include_kaiju = T,
                                            include_deepvirfinder = T)
```

## virsorter2 + checkV + kaiju
```{r}
viruses$keep_score_vs2_cv_kj <- getting_viral_set_1(viruses, include_vibrant = F,
                                            include_virsorter2 = F,
                                            include_checkV = T,
                                            include_kaiju = T,
                                            include_deepvirfinder = T)
```

# Plot of how many methods identify a contig
```{r}
viruses_high <- viruses[viruses$keep_score_all>=1,]
viruses_high_mod <- viruses_high %>% select(keep_score_vb_cv, keep_score_dvf_cv,keep_score_vs2_cv, keep_score_cv_kj)
viruses_high_mod <- apply(viruses_high_mod, c(1,2), function(x) {if (x >= 1) {x <- 1} else {x <- 0}})
viruses_high_mod <- as_tibble(viruses_high_mod)


sm <- as.data.frame(rbind(c(table(rowSums(viruses_high_mod[viruses_high_mod$keep_score_vb_cv>0,]))),
                          c(table(rowSums(viruses_high_mod[viruses_high_mod$keep_score_dvf_cv>0,]))),
                          c(table(rowSums(viruses_high_mod[viruses_high_mod$keep_score_vs2_cv>0,]))),
                          c(table(rowSums(viruses_high_mod[viruses_high_mod$keep_score_cv_kj>0,])))
                          )
)

sm$method <- c("vibrant", "deepvirfinder", "virsorter2", "kaiju")

colnames(sm) <- c("one","two","three","all","method")

sm
```

```{r}
sm_m <- melt(sm)
colnames(sm_m) <- c("primary_method", "number_methods", "number")
```

```{r}
ggplot(sm_m, aes(x=primary_method, y=number,
                   fill=number_methods)) +
  geom_bar(stat="identity") +
  theme_light() +
  coord_flip() +
  theme(
    panel.grid.major.y = element_blank(),
    panel.border = element_blank(),
    axis.ticks.y = element_blank(),
    legend.position = "bottom"
  ) +
  scale_fill_manual(name = 'Number of Methods',
                     values = alpha(c(viridis(5)), 1)) +
  xlab("Primary Method") +
  ylab("Count of Viral Contigs")
```

# Visualizing which tools led to a contig being selected
(alluvial plot)
with help from: https://www.bioinfo-scrounger.com/archives/alluvial-and-Sankey/

```{r}
alluvial_data <- viruses %>% select(keep_score_vb_cv, keep_score_dvf_cv,keep_score_vs2_cv, keep_score_cv_kj) %>% 
  apply(c(1,2), function(x) {if (x >= 1) {x <- 1} else if (x==0.5) {x <- 0.5} else {x <- 0}}) %>%
  as_tibble() %>%
  count(keep_score_vb_cv, keep_score_dvf_cv,keep_score_vs2_cv, keep_score_cv_kj)
```

```{r}
alluvial_data$cum_sum <- alluvial_data %>% select(keep_score_vb_cv, keep_score_dvf_cv,keep_score_vs2_cv, keep_score_cv_kj) %>% 
  rowSums()
```

```{r}
alluvial_data$viral <- "not viral"
alluvial_data$viral[alluvial_data$cum_sum>=1] <- "one method"
alluvial_data$viral[alluvial_data$cum_sum>=2] <- "two methods"
alluvial_data$viral[alluvial_data$cum_sum>=3] <- "three methods"
alluvial_data$viral[alluvial_data$cum_sum>=4] <- "all methods"

alluvial_data$viral <- factor(alluvial_data$viral, levels=c("not viral", "one method",
                                                            "two methods", "three methods",
                                                            "all methods"))
```

```{r}
alluvial_data <- alluvial_data %>% 
  mutate(across(c(keep_score_vb_cv, keep_score_dvf_cv,keep_score_vs2_cv, keep_score_cv_kj), as.factor))
```

```{r}
head(alluvial_data)
```


```{r}
library(ggalluvial)
```


```{r}
ggplot(data = alluvial_data,
       aes(axis1 = as.character(keep_score_vb_cv), 
           axis2 = as.character(keep_score_vs2_cv),
           axis3 = as.character(keep_score_dvf_cv),
           axis4 = as.character(keep_score_cv_kj),
           y = n)) +
  geom_alluvium(aes(fill = viral)) +
  geom_stratum() +
  geom_text(stat = "stratum",
            aes(label = after_stat(stratum))) +
  scale_fill_viridis_d() + scale_x_discrete(limits = c("VIBRANT", "VirSorter2", "DeepVirFinder", "Kaiju"),
                                  expand = c(0.05, 0.05, 0.05, 0.05)) +
  ylab("Number of Sequences") +
  theme_classic()
```




# Using NMDS to compare the tools
get matrix want to use for beta diversity
```{r}
virus_score_matrix <- viruses %>% select(keep_score_vb, keep_score_dvf,keep_score_vs2, keep_score_kj, keep_score_vb_cv, keep_score_dvf_cv,keep_score_vs2_cv, keep_score_kj_cv, keep_score_all)

```

```{r}
virus_score_matrix <- virus_score_matrix[rowSums(virus_score_matrix)>0,]
```

```{r}
library(vegan)
```


```{r}
pal <- ggthemes::tableau_color_pal(palette="Tableau 20", type="regular")


example_NMDS <- metaMDS(t(virus_score_matrix), distance = "bray")

ss <- as.data.frame(example_NMDS$points)
ss$tool <- colnames(virus_score_matrix)
```

```{r}
p <- ggplot() + 
  geom_point(data=ss, aes(x=MDS1,y=MDS2, color=tool, fill=tool, shape=tool, size=4)) +
   theme(panel.background = element_blank(), 
        axis.line = element_line(colour = "black"),
        panel.border = element_rect(colour = "black", fill=NA, size=1),
        axis.text=element_text(size=14, colour="black"),
        axis.title=element_text(size=16),
        legend.key=element_blank(),
        legend.position="right",
        legend.background = element_blank()) +
  guides(size="none") + xlab("NMDS 1") + ylab("NMDS 2") + 
  coord_equal() +
  scale_fill_manual(name="",
                     values = alpha(c("black", "#F1CE63", "#B6992D", "#8CD17D", "#59A14F", "#FFBE7D", "#F28E2B", "#A0CBE8", "#4E79A7"), 0.5)) +
  scale_color_manual(name="",
                     values = alpha(c("black","#F1CE63", "#B6992D", "#8CD17D", "#59A14F", "#FFBE7D", "#F28E2B", "#A0CBE8", "#4E79A7"), 1)) +
  scale_shape_manual(name="",
                     values = c(20,21,21,22,22,25,23,23,24,24)) 

p
```

```{r}
pal <- ggthemes::tableau_color_pal(palette="Tableau 20", type="regular")

virus_score_matrix_binary <- apply(virus_score_matrix, c(1,2), function(x) if (x>=1) {x <- 1} else {x <- 0})

example_NMDS <- metaMDS(t(virus_score_matrix_binary), distance = "bray")

ss <- as.data.frame(example_NMDS$points)
ss$tool <- colnames(virus_score_matrix)
```

```{r}
p <- ggplot() + 
  geom_point(data=ss, aes(x=MDS1,y=MDS2, color=tool, fill=tool, shape=tool, size=4)) +
   theme(panel.background = element_blank(), 
        axis.line = element_line(colour = "black"),
        panel.border = element_rect(colour = "black", fill=NA, size=1),
        axis.text=element_text(size=14, colour="black"),
        axis.title=element_text(size=16),
        legend.key=element_blank(),
        legend.position="right",
        legend.background = element_blank()) +
  guides(size="none") + xlab("NMDS 1") + ylab("NMDS 2") + 
  coord_equal() +
  scale_fill_manual(name="",
                     values = alpha(c("black", "#F1CE63", "#B6992D", "#8CD17D", "#59A14F", "#FFBE7D", "#F28E2B", "#A0CBE8", "#4E79A7"), 0.5)) +
  scale_color_manual(name="",
                     values = alpha(c("black","#F1CE63", "#B6992D", "#8CD17D", "#59A14F", "#FFBE7D", "#F28E2B", "#A0CBE8", "#4E79A7"), 1)) +
  scale_shape_manual(name="",
                     values = c(20,21,21,22,22,25,23,23,24,24)) 

p
```

```{r}
virus_score_matrix <- viruses %>% select(keep_score_vb, keep_score_dvf,keep_score_vs2, keep_score_kj, keep_score_vb_cv, keep_score_dvf_cv, keep_score_vs2_cv, keep_score_kj_cv, keep_score_vb_cv_kj, keep_score_dvf_cv_kj, keep_score_vs2_cv_kj, keep_score_vb_vs2_cv_kj, keep_score_vb_cv_kj_dvf, keep_score_vs2_cv_kj_dvf, keep_score_all)
```

```{r}
virus_score_matrix <- virus_score_matrix[rowSums(virus_score_matrix)>0,]
```

```{r}
pal <- ggthemes::tableau_color_pal(palette="Tableau 20", type="regular")

example_NMDS <- metaMDS(t(virus_score_matrix), distance = "bray")

ss <- as.data.frame(example_NMDS$points)
ss$tool <- colnames(virus_score_matrix)
```

```{r}
p <- ggplot() + 
  geom_point(data=ss, aes(x=MDS1,y=MDS2, color=tool, fill=tool, shape=tool, size=4)) +
   theme(panel.background = element_blank(), 
        axis.line = element_line(colour = "black"),
        panel.border = element_rect(colour = "black", fill=NA, size=1),
        axis.text=element_text(size=14, colour="black"),
        axis.title=element_text(size=16),
        legend.key=element_blank(),
        legend.position="right",
        legend.background = element_blank()) +
  guides(size="none") + xlab("NMDS 1") + ylab("NMDS 2") + 
  coord_equal() +
  scale_fill_manual(name="",
                     values = alpha(c("black", "#F1CE63", "#4E79A7", "#59A14F", "#8CD17D", "#4E79A7", "#FFBE7D", "#4E79A7", "#59A14F", "#F1CE63", "#E15759", "#E15759", "#4E79A7","#59A14F", "#F1CE63"), 0.5)) +
  scale_color_manual(name="",
                     values = alpha(c("black", "#F1CE63", "#4E79A7", "#59A14F", "#8CD17D", "#4E79A7", "#FFBE7D", "#4E79A7", "#59A14F", "#F1CE63", "#E15759", "#E15759", "#4E79A7","#59A14F", "#F1CE63"), 1)) +
  scale_shape_manual(name="",
                     values = c(20,21,21,21,22,22,23,23,23,23,23,24,24,24,24)) 

p
```

```{r}
virus_score_matrix <- viruses %>% select(keep_score_vb, keep_score_dvf,keep_score_vs2, keep_score_kj, keep_score_vb_cv, keep_score_dvf_cv, keep_score_vs2_cv, keep_score_kj_cv, keep_score_vb_cv_kj, keep_score_dvf_cv_kj, keep_score_vs2_cv_kj, keep_score_vb_vs2_cv_kj, keep_score_vb_cv_kj_dvf, keep_score_vs2_cv_kj_dvf, keep_score_all)
```

```{r}
virus_score_matrix <- virus_score_matrix[rowSums(virus_score_matrix)>0,]
```

```{r}
pal <- ggthemes::tableau_color_pal(palette="Tableau 20", type="regular")

virus_score_matrix_binary <- apply(virus_score_matrix, c(1,2), function(x) if (x>=1) {x <- 1} else {x <- 0})

example_NMDS <- metaMDS(t(virus_score_matrix_binary), distance = "bray")

ss <- as.data.frame(example_NMDS$points)
ss$tool <- colnames(virus_score_matrix)
```

```{r}
p <- ggplot() + 
  geom_point(data=ss, aes(x=MDS1,y=MDS2, color=tool, fill=tool, shape=tool, size=4)) +
   theme(panel.background = element_blank(), 
        axis.line = element_line(colour = "black"),
        panel.border = element_rect(colour = "black", fill=NA, size=1),
        axis.text=element_text(size=14, colour="black"),
        axis.title=element_text(size=16),
        legend.key=element_blank(),
        legend.position="right",
        legend.background = element_blank()) +
  guides(size="none") + xlab("NMDS 1") + ylab("NMDS 2") + 
  coord_equal() +
  scale_fill_manual(name="",
                     values = alpha(c("black", "#F1CE63", "#4E79A7", "#59A14F", "#8CD17D", "#4E79A7", "#FFBE7D", "#4E79A7", "#59A14F", "#F1CE63", "#E15759", "#E15759", "#4E79A7","#59A14F", "#F1CE63"), 0.5)) +
  scale_color_manual(name="",
                     values = alpha(c("black", "#F1CE63", "#4E79A7", "#59A14F", "#8CD17D", "#4E79A7", "#FFBE7D", "#4E79A7", "#59A14F", "#F1CE63", "#E15759", "#E15759", "#4E79A7","#59A14F", "#F1CE63"), 1)) +
  scale_shape_manual(name="",
                     values = c(20,21,21,21,22,22,23,23,23,23,23,24,24,24,24)) 

p
```
