---
title: "Identifying Viral Contigs"
output: html_notebook
---

```{r}
library(ggplot2)
library(plyr)
library(reshape2)
library(viridis)
library(treemapify)
library(tidyr)
library(pheatmap)
library(dplyr)
library(gridExtra)
library(readr)
```



# All Viral Contigs

## checkV
```{r}
checkV <- read_tsv("../../ToolOutput/quality_summary.tsv")
checkV$contig <- sub("\\.", "_", checkV$contig_id)
```

## VIBRANT
```{r}
vb_c <- read_tsv("../../ToolOutput/VIBRANT_genome_quality_combined_virus_refseq.tsv")
vb_c$contig <- sub("\\.", "_", vb_c$scaffold)
vb_c$contig <- gsub("_fragment_.*", "", vb_c$contig)
vb_c <- separate(vb_c, col=contig, into="contig", remove=T, sep = " ")
vb_c <- vb_c[!duplicated(vb_c$contig),]
```

## DeepVirFinder
```{r}
dvf_c <- read_tsv("../../ToolOutput/combined_virus_refseq_2000000.fasta_gt1bp_dvfpred.txt")
dvf_c$contig <- sub("\\.", "_", dvf_c$name)
dvf_c$bh_pvalue <- p.adjust(dvf_c$pvalue, method="BH")
```

## VirSorter2 
```{r}
vs2_c <- read_tsv("../../ToolOutput/final-viral-score.tsv")
vs2_c <- separate(vs2_c, col = seqname, into = c("Contig", "Type"), "\\|\\|")
vs2_c$contig <- sub("\\.", "_", vs2_c$Contig)
vs2_c <- vs2_c[!duplicated(vs2_c$contig),]
```

## Merging
```{r}
viruses <- full_join(x=checkV, y=dvf_c, by = c("contig"))
viruses <- full_join(x=viruses, y=vb_c, by = c("contig"))
viruses <- full_join(x=viruses, y=vs2_c, by = c("contig"))
```

```{r}
head(viruses)
```



```{r}
viruses$percent_host <- viruses$host_genes/viruses$gene_count*100
viruses$percent_viral <- viruses$viral_genes/viruses$gene_count*100
viruses$percent_unknown <- 100-(viruses$host_genes+viruses$viral_genes)/viruses$gene_count*100

viruses$Quality[is.na(viruses$Quality)] <- 0
viruses$score[is.na(viruses$score)] <- 0
viruses$viral[is.na(viruses$viral)] <- 0
viruses$hallmark[is.na(viruses$hallmark)] <- 0
viruses$percent_viral[is.na(viruses$percent_viral)] <- 0
```


######################################################################
# New rules for keeping

## Getting Score for Deciding Whether to Keep
```{r}
viruses$keep_score_vibrant <- 0
viruses$keep_score_deepvirfinder <- 0
viruses$keep_score_virsorter2 <- 0
viruses$keep_score_checkV <- 0

#vibrant
viruses$keep_score_vibrant[viruses$Quality=="high quality draft"] <- 1
viruses$keep_score_vibrant[viruses$Quality=="medium quality draft"] <- 0.52
viruses$keep_score_vibrant[viruses$Quality=="low quality draft"] <- 0.333
viruses$keep_score_vibrant[viruses$viral_genes==0 & viruses$host_genes>=1] <- 0
viruses$keep_score_vibrant[(viruses$viral_genes*3) <= viruses$host_genes] <- 0
viruses$keep_score_vibrant[viruses$contig_length>50000 & viruses$hallmark==0] <- 0

#deepvirfinder
viruses$keep_score_deepvirfinder[viruses$bh_pvalue<=0.05] <- 1
viruses$keep_score_deepvirfinder[viruses$viral_genes==0 & viruses$host_genes>=1] <- 0
viruses$keep_score_deepvirfinder[(viruses$viral_genes*3) <= viruses$host_genes] <- 0
viruses$keep_score_deepvirfinder[viruses$contig_length>50000 & viruses$hallmark==0] <- 0

#virsorter2
viruses$keep_score_virsorter2[viruses$viral>=50] <- 0.333
viruses$keep_score_virsorter2[viruses$hallmark>2] <- 1
viruses$keep_score_virsorter2[viruses$viral>=95] <- viruses$keep_score_virsorter2[viruses$viral>=95] + 1
viruses$keep_score_virsorter2[viruses$viral_genes==0 & viruses$host_genes>=1] <- 0
viruses$keep_score_virsorter2[(viruses$viral_genes*3) <= viruses$host_genes] <- 0
viruses$keep_score_virsorter2[viruses$contig_length>50000 & viruses$hallmark==0] <- 0

#checkV
viruses$keep_score_checkV[(viruses$viral>=50 & viruses$host_genes==0) & viruses$viral_genes==0] <- viruses$keep_score_checkV[(viruses$viral>=50 & viruses$host_genes==0) & viruses$viral_genes==0] + 0.52
viruses$keep_score_checkV[viruses$percent_viral>=50] <- viruses$keep_score_checkV[viruses$percent_viral>=50] + 1
viruses$keep_score_checkV[viruses$viral_genes==0 & viruses$host_genes>=1] <- 0
viruses$keep_score_checkV[(viruses$viral_genes*3) <= viruses$host_genes] <- 0
viruses$keep_score_checkV[viruses$contig_length>50000 & viruses$hallmark==0] <- 0
```

### keep high quality by any method (minus discard rules at bottom), medium quality in at least 3, low if consensus with others 
```{r}
viruses$keep_score <- viruses$keep_score_checkV + viruses$keep_score_virsorter2 +
  viruses$keep_score_deepvirfinder + viruses$keep_score_vibrant

viruses$keep_score_high <- 0
viruses$keep_score_high[viruses$keep_score>=1] <- 1
```

```{r}
table(viruses$keep_score_high)
```


```{r}
viruses_high <- viruses[viruses$keep_score_high==1,]
```

```{r}
head(viruses_high)
```


# Plot of how many methods identify a contig
```{r}
viruses_high_mod <- viruses_high %>% select("keep_score_vibrant":"keep_score_checkV")
viruses_high_mod <- apply(viruses_high_mod, c(1,2), function(x) {if (x > 0) {x <- 1} else {x <- 0}})
viruses_high_mod <- as_tibble(viruses_high_mod)


sm <- as.data.frame(rbind(c(0, table(rowSums(viruses_high_mod[viruses_high_mod$keep_score_vibrant>0,]))),
                          table(rowSums(viruses_high_mod[viruses_high_mod$keep_score_deepvirfinder>0,])),
                          table(rowSums(viruses_high_mod[viruses_high_mod$keep_score_checkV>0,])),
                          table(rowSums(viruses_high_mod[viruses_high_mod$keep_score_virsorter2>0,]))
                          )
)

sm$method <- c("vibrant", "deepvirfinder", "checkV_polishing", "virsorter2")

colnames(sm) <- c("one","two","three","four", "method")

sm
```

```{r}
sm_m <- melt(sm)
colnames(sm_m) <- c("primary_method", "number_methods", "number")
```

```{r}
ggplot(sm_m, aes(x=primary_method, y=number,
                   fill=number_methods)) +
  geom_bar(stat="identity") +
  theme_light() +
  coord_flip() +
  theme(
    panel.grid.major.y = element_blank(),
    panel.border = element_blank(),
    axis.ticks.y = element_blank(),
    legend.position = "bottom"
  ) +
  scale_fill_manual(name = 'Number of Methods',
                     values = alpha(c(viridis(7)), 1)) +
  xlab("Primary Method") +
  ylab("Count of Viral Contigs")
```


```{r}
hist(viruses$keep_score[viruses$keep_score_high==1])
```


```{r}


ggplot(viruses_high, aes(x=percent_host, y=percent_viral,
                   color=keep_score, fill=keep_score)) +
  geom_point(alpha=0.5) +
  theme_light() +
  theme(
    panel.grid.major.y = element_blank(),
    panel.border = element_blank(),
    axis.ticks.y = element_blank(),
    legend.position = "bottom"
  ) +
  scale_fill_viridis(option = "viridis") +
  scale_color_viridis(option = "viridis") +
  xlab("Percent Genes Host") +
  ylab("Percent Genes Viral") +
  coord_flip()

ggplot(viruses_high, aes(x=percent_host, y=percent_viral,
                   color=score, fill=score)) +
  geom_point(alpha=0.5) +
  theme_light() +
  theme(
    panel.grid.major.y = element_blank(),
    panel.border = element_blank(),
    axis.ticks.y = element_blank(),
    legend.position = "bottom"
  ) +
  scale_fill_viridis(option = "magma") +
  scale_color_viridis(option = "magma") +
  xlab("Percent Genes Host") +
  ylab("Percent Genes Viral") +
  coord_flip()


ggplot(viruses_high, aes(x=percent_host, y=percent_viral,
                   color=log10(0.1+hallmark), 
                   fill=log10(0.1+hallmark))) +
  geom_point(alpha=0.5) +
  theme_light() +
  theme(
    panel.grid.major.y = element_blank(),
    panel.border = element_blank(),
    axis.ticks.y = element_blank(),
    legend.position = "bottom"
  ) +
  scale_fill_viridis(option = "viridis") +
  scale_color_viridis(option = "viridis") +
  xlab("Percent Genes Host") +
  ylab("Percent Genes Viral") +
  coord_flip()

ggplot(viruses_high, aes(x=percent_host, y=percent_viral,
                   color=log10(contig_length), fill=log10(contig_length))) +
  geom_point(alpha=0.5) +
  theme_light() +
  theme(
    panel.grid.major.y = element_blank(),
    panel.border = element_blank(),
    axis.ticks.y = element_blank(),
    legend.position = "bottom"
  ) +
  scale_fill_viridis(option = "magma") +
  scale_color_viridis(option = "magma") +
  xlab("Percent Genes Host") +
  ylab("Percent Genes Viral") +
  coord_flip()

ggplot(viruses_high, aes(x=percent_viral, y=contig_length,
                   color=log10(0.1+hallmark), 
                   fill=log10(0.1+hallmark))) +
  geom_point(alpha=0.1) +
  theme_light() +
  theme(
    panel.grid.major.y = element_blank(),
    panel.border = element_blank(),
    axis.ticks.y = element_blank(),
    legend.position = "bottom"
  ) +
  scale_fill_viridis(option = "viridis") +
  scale_color_viridis(option = "viridis") +
  xlab("Percent Viral") +
  ylab("Contig Length") +
  scale_y_log10()

ggplot(viruses_high, aes(x=percent_viral, y=percent_unknown,
                   color=log10(contig_length), 
                   fill=log10(contig_length))) +
  geom_point(alpha=0.1) +
  theme_light() +
  theme(
    panel.grid.major.y = element_blank(),
    panel.border = element_blank(),
    axis.ticks.y = element_blank(),
    legend.position = "bottom"
  ) +
  scale_fill_viridis(option = "magma") +
  scale_color_viridis(option = "magma") +
  xlab("Percent Viral") +
  ylab("Percent Unknown") 
```



