---
title: "Building Environmental Data Viral Features Set Drinking Water"
author: Bridget Hegarty, James Riddell
date: 10-26-2022
output: html_notebook
---
This Rmarkdown file takes outputs from viral identification tools and merges
them into one file, "{dataset}_merged_viral_id_outputs.csv", which is the input
for identifying_viral_metrics_environmental.Rmd.

Please reach out to James Riddell (riddell.26@buckeyemail.osu.edu) or
Bridget Hegarty (beh53@case.edu) regarding any issues, or open an issue on github.


```{r setup-library}
library(ggplot2)
library(plyr)
library(reshape2)
library(viridis)
library(tidyr)
library(dplyr)
library(readr)
library(data.table)
```

# Inputs
1) For this script's inputs, the user must merge each viral identification 
tool output with the following specified filenames into one tab-separated file. 
Replace the filenames with their paths if needed into the variables in the 
'inputs' chunk.

    General filenames for each tool:
        Checkv: quality_summary.tsv
        VIBRANT: VIBRANT_genome_quality_${assembly}.tsv
        Virsorter: VIRSorter_global-phage-signal.csv
        Virsorter2: final-viral-score.tsv
        DeepVirFinder: ${assembly}.fasta_gt2500bp_dvfpred.txt
        Kaiju: ${assembly}.kaiju.names.out

2) Make sure all fasta headers (assembly, contig) are consistent within and
across tools. Chunks for each tool do contain some lines for cleaning up these features, but due to their variability it will be the user's responsibility to 
make sure they match across tools.

3) Check each chunk and ensure all columns are accounted for. For example,
if you don't have "assembly" and called it something else, please either
remove it from this script or modify your input files.

4) This script is designed for contigs > 3000 basepairs. It can be modified
to be higher or lower, but going lower will greatly increase the size of the
dataframe and memory usage.



```{r inputs}
# dataset name for file organization and outputs for metrics file
dataset_name <- "drinking_water"
# checkV
checkV_path <- "DrinkingWaterToolOutput/merged_checkV_allcontigs.tsv"
# VIBRANT
vibrant_path <- "DrinkingWaterToolOutput/vibrant_merged_contig_list_all.txt"
# DeepVirFinder
dvf_path <- "DrinkingWaterToolOutput/dvf_outputs_merged.tsv"
# Virsorter
vs_path <- "DrinkingWaterToolOutput/vs_merged_all.txt"
# Virsorter2
vs2_path <- "DrinkingWaterToolOutput/vs2_merged_file.txt"
# Kaiju
kj_path <- "/Users/hegartyb/Documents/BlueSkyPostdoc/ViralSortingToolsEvaluation/Temp/kaiju_outputs_merged.tsv"
# KB cutoff
KB_CUTOFF <- 3000

```

# All Viral Contigs

## checkV
```{r}
checkV <- fread(checkV_path, 
                sep="\t",
                header = T, 
                select = c(
                    'contig_id',
                    'assembly',
                    'provirus',
                    'completeness',
                    'contamination',
                    'viral_genes',
                    'host_genes',
                    'gene_count',
                    'contig_length',
                    'checkv_quality'
                    )
                ) %>% 
    rename(
        contig = contig_id,
        checkv_provirus = provirus,
        checkv_completeness = completeness,
        checkv_contamination = contamination,
        checkv_viral_genes = viral_genes,
        checkv_host_genes = host_genes,
        checkv_total_genes = gene_count,
        checkv_length = contig_length
        )
checkV$method = 'checkv'
checkV$contig <- sub("\\.", "_", checkV$contig)
checkV$uniq_contig <- paste(checkV$assembly, checkV$contig, sep = "--")
checkV <- checkV[!duplicated(checkV$uniq_contig),]
```

## VIBRANT
```{r}
vb_c <- fread(vibrant_path,
              header = T,
              sep = "\t",
              select = c(
                  'scaffold',
                  'assembly',
                  'type',
                  'quality'
                  )
              ) %>%
    rename(
        contig = scaffold,
        vibrant_quality = quality
        )
vb_c$method <- "vibrant"
vb_c$vibrant_prophage <- "No"
vb_c$vibrant_prophage[grep("_fragment_", vb_c$contig)] <- "Yes"
vb_c$contig <- gsub("_fragment_.*", "", vb_c$contig)
vb_c$contig <- sub("\\.", "_", vb_c$contig)
vb_c$uniq_contig <- paste(vb_c$assembly, vb_c$contig, sep = "--")
vb_c <- vb_c[!duplicated(vb_c$uniq_contig),]
```

## DeepVirFinder
```{r}
dvf_c <- fread(dvf_path,
               header = T,
               sep = "\t",
               select = c(
                   'sample',
                   'name',
                   'score',
                   'pvalue'
                   )
               ) %>% 
    rename(
        contig = name,
        assembly = sample
        )
dvf_c$contig <- sub("\\.", "_", dvf_c$contig)
dvf_c$bh_pvalue <- p.adjust(dvf_c$pvalue, method = "BH")
dvf_c$uniq_contig <- paste(dvf_c$assembly, dvf_c$contig, sep = "--")
dvf_c <- dvf_c[!duplicated(dvf_c$uniq_contig),]
```

## VirSorter
```{r}
vs_c <- fread(vs_path,
              select = c(
                  'assembly',
                  'contig_id',
                  'category'
                  ),
              ) %>% 
  rename(contig = contig_id) %>%
  drop_na(contig)
vs_c$contig <- sub("VIRSorter_", "", vs_c$contig)
vs_c$contig <- sub("-circula", "", vs_c$contig)
vs_c$contig <- sub("uth", "South", vs_c$contig)
vs_c$contig <- sub("pannenburg", "Spannenburg", vs_c$contig)
vs_c$contig <- sub("\\.", "_", vs_c$contig)
vs_c$uniq_contig <- paste(vs_c$assembly, vs_c$contig, sep = "--")

vs_c <- vs_c[!duplicated(vs_c$uniq_contig),]
vs_c <- vs_c %>% drop_na(uniq_contig)
```


## VirSorter2 
```{r}
vs2_c <- fread(vs2_path,
                    header = T,
                    sep = '\t',
                    select = c(
                        'seqname',
                        'assembly',
                        'dsDNAphage',
                        'ssDNA',
                        'max_score',
                        'max_score_group',
                        'hallmark',
                        'viral',
                        'cellular'
                    )
               ) %>% 
              separate(
                  col = seqname,
                  into = c("contig", "type"), 
                  sep = "\\|\\|",
                  remove = T
                  )
vs2_c$assembly <- sub("_3000", "", vs2_c$assembly)
vs2_c$contig <- sub("\\.", "_", vs2_c$contig)
vs2_c$uniq_contig <- paste(vs2_c$assembly, vs2_c$contig, sep="--")
vs2_c <- vs2_c[!duplicated(vs2_c$uniq_contig),]
```


## Kaiju
```{r}
kj_c <- read_tsv(kj_path, col_names = T)
colnames(kj_c)[1] <- "assembly"
colnames(kj_c)[3] <- "contig"
kj_c$contig <- sub("\\.", "_", kj_c$contig)
kj_c <- separate(kj_c, col = Name, into = c("Kaiju_Viral","Kingdom"), sep=";")
kj_c$uniq_contig <- paste(kj_c$assembly, kj_c$contig, sep="--")
kj_c <- kj_c[!duplicated(kj_c$contig),]
```

```{r}
# Quick glimpse of cellular vs viral sequence classification by Kaiju by assembly
table(kj_c$assembly, kj_c$Kaiju_Viral)
```


## Merging
```{r}
viruses <- full_join(x=vs2_c, y=checkV, by = c("uniq_contig", "assembly"))
viruses <- full_join(x=viruses, y=dvf_c, by = c("uniq_contig", "assembly"))
viruses <- full_join(x=viruses, y=vb_c, by = c("uniq_contig", "assembly"))
viruses <- full_join(x=viruses, y=vs_c, by = c("uniq_contig", "assembly"))
viruses <- full_join(x=viruses, y=kj_c, by = c("uniq_contig", "assembly"))
```

Remove contigs not greater than the basepair length cutoff (3000)

```{r}
viruses <- viruses %>% filter(checkv_length > KB_CUTOFF)
```


Use the next code chunk to check if any contigs are missing from checkv.
If yes, then there is a mix up in contig names since checkv should contain
all contig names even if they are identified as non-viral.
```{r}
v_missing <- viruses[is.na(viruses$checkv_uniq_contig),]
```



calculate percent viral/host/unknown
```{r}
viruses$percent_host <- viruses$checkv_host_genes/viruses$checkv_total_genes*100
viruses$percent_viral <- viruses$checkv_viral_genes/viruses$checkv_total_genes*100
viruses$percent_unknown <- 100-(viruses$checkv_host_genes+viruses$checkv_viral_genes)/viruses$checkv_total_genes*100
```

get rid of NAs for downstream processing
```{r}

# checkV
viruses$percent_viral[is.na(viruses$percent_viral)] <- 0
viruses$percent_unknown[is.na(viruses$percent_unknown)] <- 0
viruses$checkv_completeness[is.na(viruses$checkv_completeness)] <- 0

# VIBRANT
viruses$vibrant_quality[is.na(viruses$vibrant_quality)] <- 0

# DeepVirFinder
viruses$score[is.na(viruses$score)] <- 0
viruses$bh_pvalue[is.na(viruses$bh_pvalue)] <- 0

# VirSorter2
viruses$viral[is.na(viruses$viral)] <- 0
viruses$hallmark[is.na(viruses$hallmark)] <- 0

#Virsorter
viruses$category[is.na(viruses$category)] <- 0

# Kaiju
viruses$Kaiju_Viral[is.na(viruses$Kaiju_Viral)] <- "unknown"
```

```{r}
viruses_filename <- 'merged_viral_feature_outputs.csv'
write.csv(viruses, paste(dataset_name, viruses_filename, sep='_'))
```

Data cleaning is done! Export viruses to tsv and load into identifying_viral_contigs_environmental.Rmd for metrics on tool outputs. The code for metrics is also pasted below, but it is not recommended running it in the same R session due to memory issues. The construction of the final dataframe essentially doubles the memory being used, and even with garbage collector can make Rstudio unresponsive.